File: suites/13_simd_vectorization/comparisons/nbody_simulation.go
package main

import "fmt"

func nbodySimulation(bodies int64) int64 {
	steps := int64(10)
	checksum := int64(0)

	for step := int64(0); step < steps; step++ {
		for i := int64(0); i < bodies; i++ {
			px := ((i*17 + step) % 2000) - 1000
			py := ((i*19 + step) % 2000) - 1000
			pz := ((i*23 + step) % 2000) - 1000

			fx, fy, fz := int64(0), int64(0), int64(0)
			for j := int64(0); j < bodies; j++ {
				if i != j {
					ox := ((j*17 + step) % 2000) - 1000
					oy := ((j*19 + step) % 2000) - 1000
					oz := ((j*23 + step) % 2000) - 1000

					dx, dy, dz := ox-px, oy-py, oz-pz
					dSq := dx*dx + dy*dy + dz*dz
					if dSq == 0 { dSq = 1 }

					invDistCubed := 1000000 / (dSq * dSq)
					fx += dx * invDistCubed
					fy += dy * invDistCubed
					fz += dz * invDistCubed
				}
			}
			checksum = (checksum*31 + fx + fy + fz) % 1000000007
		}
	}
	return checksum
}

func main() { fmt.Println(nbodySimulation(256)) }

File: suites/13_simd_vectorization/comparisons/nbody_simulation.rs
fn nbody_simulation(bodies: i64) -> i64 {
    let steps: i64 = 10;
    let mut checksum: i64 = 0;
    
    for step in 0..steps {
        for i in 0..bodies {
            let px = ((i * 17 + step) % 2000) - 1000;
            let py = ((i * 19 + step) % 2000) - 1000;
            let pz = ((i * 23 + step) % 2000) - 1000;
            
            let mut fx: i64 = 0; let mut fy: i64 = 0; let mut fz: i64 = 0;
            for j in 0..bodies {
                if i != j {
                    let ox = ((j * 17 + step) % 2000) - 1000;
                    let oy = ((j * 19 + step) % 2000) - 1000;
                    let oz = ((j * 23 + step) % 2000) - 1000;
                    
                    let dx = ox - px; let dy = oy - py; let dz = oz - pz;
                    let mut d_sq = dx*dx + dy*dy + dz*dz;
                    if d_sq == 0 { d_sq = 1; }
                    
                    let inv_dist_cubed = 1000000 / (d_sq * d_sq);
                    fx += dx * inv_dist_cubed;
                    fy += dy * inv_dist_cubed;
                    fz += dz * inv_dist_cubed;
                }
            }
            checksum = (checksum * 31 + fx + fy + fz) % 1000000007;
        }
    }
    checksum
}

fn main() { println!("{}", nbody_simulation(256)); }
